keps/sig-api-machinery/20190603-immutable-fields.md

```diff
@@ -0,0 +1,204 @@
+see-also:
+  - "/keps/sig-api-machinery/0006-apply.md"
+---
+
+# Immutable fields
+
+## Release Signoff Checklist
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+A lot of fields in APIs tend to be "immutable", they can't be changed after
+creation. This is true for example for many of the fields in pods. There is
+currently no way to declaratively document that fields are immutable, and one
+has to rely on either built-in validation for core types, or have to build a
+validating webhooks for CRDs.
+
+Providing a new `// +immutable` marker would make the API more descriptive to
+users and help API developers by verifying these assertions automatically.
+
+## Motivation
+
+There are resources in Kubernetes which have immutable fields by design,
+i.e. after creation of an object, those fields cannot be mutated anymore. E.g. a
+pod's specification is mostly unchangeable once it is created. To change the
+pod, it must be deleted, recreated and rescheduled. Users want to implement the
+same kind of read-only semantics for CustomResources, for example:
+https://github.com/kubernetes/kubernetes/issues/65973
+
+### Goals
+
+The goals are mostly twofold:
+- Provide a mechanism to document APIs (both built-in types and CRDs) by having
+  that additional marker provide information to users AND tools (including
+  generated documentation, kubectl explain),
+- Automatically provide logic for developers (of core-types and CRDs) to
+  guarantee immutability of specific fields.
+
+### Non-Goals
+
+There are a few non-goals:
+- OpenAPI has a notion of "readOnly", but this is only meant to simplify schema,
+  and one is never supposed to send a readOnly value. This is different because
+  we want "writeOnce", not "readOnly".
+
+## Proposal
+
+This proposal attempts to create a concept of immutable and to visit some of
+what this can mean, but doesn't intend to propose all the possible
+semantics. The proposal should not close the door to further improvements.
+
+### Semantics
+
+We'll define immutable as "writeOnce", which means that the fields can only be
+set at creation time, and can never be updated after. Attempts to update an
+immutable field will result in an error (as opposed to being ignored), though we
+could potentially add that semantics later.
```

Comment:
Perhaps add the following, to make it explicit:

---
Clients implement  [idempotent creation](https://kubernetes.io/docs/reference/using-api/api-concepts/#standard-api-terminology) by retrying creation, and they currently do not get errors if the same creation is tried twice.  This KEP will not change that.   Specifically, two consecutive POSTs of the same object, which are mutated the same way by mutating admission controller (which is the typical case), will both succeed.  If an error were returned the second time, and the client did not see the first success response, the client would be confused.